### PR TITLE
feat(socket sink): Add unix domain socket support to the `socket` sink

### PR DESCRIPTION
**Pull request by stbrody**
Signed-off-by: Spencer T Brody <spencer.t.brody@gmail.com>

### CODE REVIEW

**Comment by binarylogic** [comment]
@bruce, would you mind reviewing this?

File: src/sinks/util/unix.rs [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+            .expect("Unable to encode event as JSON."),
+        Encoding::Text => log
+            .get(&event::MESSAGE)
+            .map(|v| v.as_bytes())
+            .unwrap_or_default(),
+    };
+
+    // Add newline
+    let newlinebytes = b"\n";
+    bytes.extend(newlinebytes.iter());
```
Comment (bruceg) [inline src/sinks/util/unix.rs]: A simple `bytes.extend(b"\n")` works here.

File: src/sinks/util/unix.rs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    let check = future::lazy(move || {
+        UnixStream::connect(&path)
+            .map(|_| ())
+            .map_err(|source| HealthcheckError::ConnectError { source }.into())
+    });
+
+    Box::new(check)
+}
+
+fn encode_event(event: Event, encoding: &Encoding) -> Bytes {
```
Comment (bruceg) [inline src/sinks/util/unix.rs]: I notice this works almost the same as the parallel `encode_event` in `src/sinks/util/tcp.rs` but not identical. Is there a reason for the differences? Could these be unified?

File: src/sinks/util/unix.rs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    let log = event.into_log();
+
+    let mut bytes = match encoding {
+        Encoding::Json => serde_json::to_vec(&log.unflatten())
+            .map(Bytes::from)
+            .expect("Unable to encode event as JSON."),
```
Comment (bruceg) [inline src/sinks/util/unix.rs]: If the JSON encoding fails, we definitely want an error, but not to panic Vector. This should either return an error or display it and return `None` like the other `encode_event` does.

**Comment by bruceg** [comment]
Most of `src/sinks/util/unix.rs` is a straight clone of `src/sinks/util/tcp.rs` except for the difference in address scheme (path vs network) and an extra state (`TlsConnecting` in `tcp.rs`). That's seems like a lot of duplication for a very small difference. Is there any sense in making just the address family distinct rather than the whole source? Are there other opportunities for merging, like a generic `struct StreamSink<A>` where `A` is the address family?

File: .meta/sinks/socket.toml [hunk 4]
```diff
@@ -8,21 +8,32 @@ encodings = ["json", "text"]
...
+relevant_when = {mode = ["tcp"]}
 description = "The address to bind to."
+
+[sinks.socket.options.path]
+type = "string"
+common = true
+examples = ["/path/to/socket"]
+required = true
+relevant_when = {mode = "unix"}
+description = """The unix socket path. *This should be absolute path*.\
```
[the remark this answers is not in the document]
Comment (stbrody) [reply .meta/sinks/socket.toml]: This was copied from the syslog source, but I agree with your suggestions, so I updated the wording here.  Left the existing usages alone for now though.

File: .meta/sinks/socket.toml [hunk 5]
```diff
@@ -8,21 +8,32 @@ encodings = ["json", "text"]
 function_category = "transmit"
 healthcheck = true
 input_types = ["log"]
-write_to_description = "a socket, such as a TCP, UDP, or Unix socket"
+write_to_description = "a socket, such as a TCP or Unix Domain Socket"
```
[the remark this answers is not in the document]
Comment (stbrody) [reply .meta/sinks/socket.toml]: Done

File: src/sinks/http.rs [hunk 6]
```diff
@@ -501,7 +501,6 @@ mod tests {
 
         let (input_lines, events) = random_lines_with_stream(100, num_lines);
         let pump = sink.send_all(events);
-
```
[the remark this answers is not in the document]
Comment (stbrody) [reply src/sinks/http.rs]: Whoops, not sure how that snuck in there.  Good catch.

File: src/sinks/util/unix.rs (hunk 1 above)
Comment (stbrody) [reply src/sinks/util/unix.rs]: This code has been removed in favor of the impl from tcp.rs

File: src/sinks/util/unix.rs (hunk 2 above)
Comment (stbrody) [reply src/sinks/util/unix.rs]: It looks like there are currently 14 different implementations of 'encode_event' in the vector source that are all similar but slightly different, which is why I didn't bother unifying these two usages.  These two are definitely similar enough to be unified pretty easily though, so I moved the existing impl from tcp.rs into sinks/util/mod.rs and use it in both tcp.rs and unix.rs.  Hopefully over time more of the other encode_event impls can be merged into this new common one.

File: src/sinks/util/unix.rs (hunk 3 above)
Comment (stbrody) [reply src/sinks/util/unix.rs]: Good point.  Switching to the existing impl from tcp.rs fixes this

**Comment by stbrody** [comment]
> Most of `src/sinks/util/unix.rs` is a straight clone of `src/sinks/util/tcp.rs` except for the difference in address scheme (path vs network) and an extra state (`TlsConnecting` in `tcp.rs`). That's seems like a lot of duplication for a very small difference. Is there any sense in making just the address family distinct rather than the whole source? Are there other opportunities for merging, like a generic `struct StreamSink<A>` where `A` is the address family?

I like this idea, the current duplication is definitely not ideal.  The TLS support in the TCP impl might make this a little trickier, but I'm sure could be handled.  Doing this now though would increase the scope of this work, and I made the decision to leave the impls separate based on a conversion with @lukesteensen where he suggested that at this point in vector development code duplication is preferable to prematurely introducing the wrong abstraction layers.  So I'm going to defer to Luke and @binarylogic on whether to do this now or hold off for a later cleanup pass.

File: src/sinks/util/unix.rs (hunk 2 above)
Comment (bruceg) [reply src/sinks/util/unix.rs]: Looks good

**Pull request merged**